doc/yabai.1
'\" t
.\"     Title: yabai
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.10
.\"      Date: 2020-09-16
.\"    Manual: Yabai Manual
.\"    Source: Yabai
.\"  Language: English
.\"
.TH "YABAI" "1" "2020-09-16" "Yabai" "Yabai Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
.sp
yabai
.SH "SYNOPSIS"
.sp
\fByabai\fP [\fB\-v\fP,\fB\-\-version\fP|\fB\-V\fP,\fB\-\-verbose\fP|\fB\-m\fP,\fB\-\-message\fP \fImsg\fP|\fB\-c\fP,\fB\-\-config\fP \fIconfig_file\fP|\fB\-\-install\-sa\fP|\fB\-\-uninstall\-sa\fP|\fB\-\-check\-sa\fP|\fB\-\-load\-sa\fP]
.SH "DESCRIPTION"
.sp
\fByabai\fP is a tiling window manager for macOS based on binary space partitioning.
.SH "OPTIONS"
.sp
\fB\-v\fP, \fB\-\-version\fP
.RS 4
Print the version and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-verbose\fP
.RS 4
Output debug information to stdout.
.RE
.sp
\fB\-m\fP, \fB\-\-message\fP \fI<msg>\fP
.RS 4
Send message to a running instance of yabai.
.RE
.sp
\fB\-c\fP, \fB\-\-config\fP \fI<config_file>\fP
.RS 4
Use the specified configuration file.
.RE
.sp
\fB\-\-install\-sa\fP
.RS 4
Install scripting\-addition. Must be run as root.
.br
Path is /System/Library/ScriptingAdditions on macOS High Sierra, and /Library/ScriptingAdditions on Mojave and newer.
.RE
.sp
\fB\-\-uninstall\-sa\fP
.RS 4
Uninstall scripting\-addition. Must be run as root.
.br
Path is /System/Library/ScriptingAdditions on macOS High Sierra, and /Library/ScriptingAdditions on Mojave and newer.
.RE
.sp
\fB\-\-check\-sa\fP
.RS 4
Returns a zero exit\-code if the latest version of the scripting\-addition is installed.
.RE
.sp
\fB\-\-load\-sa\fP
.RS 4
Loads the scripting\-addition into Dock.app.
.RE
.SH "DEFINITIONS"
.sp
.if n .RS 4
.nf
REGEX       := \c
.URL "https://www.gnu.org/software/findutils/manual/html_node/find_html/posix_002dextended\-regular\-expression\-syntax.html" "POSIX extended regular expression syntax" ""

LABEL       := arbitrary string/text used as an identifier

LAYER       := below | normal | above

BOOL_SEL    := on | off

FLOAT_SEL   := 0 < <value> <= 1.0

RULE_SEL    := <index> | LABEL

SIGNAL_SEL  := <index> | LABEL

DIR_SEL     := north | east | south | west

STACK_SEL   := stack.prev | stack.next | stack.first | stack.last | stack.recent

WINDOW_SEL  := prev | next | first | last | recent | mouse | largest | smallest | STACK_SEL | DIR_SEL | <window id>

DISPLAY_SEL := prev | next | first | last | recent | mouse | DIR_SEL | <arrangement index (1\-based)>

SPACE_SEL   := prev | next | first | last | recent | mouse | <mission\-control index (1\-based)> | LABEL
.fi
.if n .RE
.SH "DOMAINS"
.SS "Config"
.SS "General Syntax"
.sp
yabai \-m config <global setting>
.RS 4
Get or set the value of <global setting>.
.RE
.sp
yabai \-m config [\-\-space \fI<mission\-control index>\fP] <space setting>
.RS 4
Get or set the value of <space setting>.
.RE
.SS "Global Settings"
.sp
\fBdebug_output\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable output of debug information to stdout.
.RE
.sp
\fBexternal_bar\fP [\fI<main|all|off>:<top_padding>:<bottom_padding>\fP]
.RS 4
Specify top and bottom padding for a potential custom bar that you may be running.
.br
\fImain\fP: Apply the given padding only to spaces located on the main display.
.br
\fIall\fP:  Apply the given padding to all spaces regardless of their display.
.br
\fIoff\fP:  Do not apply any special padding.
.RE
.sp
\fBmouse_follows_focus\fP [\fI<BOOL_SEL>\fP]
.RS 4
When focusing a window, put the mouse at its center.
.RE
.sp
\fBfocus_follows_mouse\fP [\fIautofocus|autoraise|off\fP]
.RS 4
Automatically focus the window under the mouse.
.RE
.sp
\fBwindow_placement\fP [\fIfirst_child|second_child\fP]
.RS 4
Specify whether managed windows should become the first or second leaf\-node.
.RE
.sp
\fBwindow_topmost\fP [\fI<BOOL_SEL>\fP]
.RS 4
Make floating windows stay on top.
.RE
.sp
\fBwindow_shadow\fP [\fI<BOOL_SEL>|float\fP]
.RS 4
Draw shadow for windows.
.RE
.sp
\fBwindow_opacity\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable opacity for windows.
.RE
.sp
\fBwindow_opacity_duration\fP [\fI<floating point number>\fP]
.RS 4
Duration of transition between active / normal opacity.
.RE
.sp
\fBactive_window_opacity\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Opacity of the focused window.
.RE
.sp
\fBnormal_window_opacity\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Opacity of an unfocused window.
.RE
.sp
\fBwindow_border\fP [\fI<BOOL_SEL>\fP]
.RS 4
Draw border for windows.
.RE
.sp
\fBwindow_border_width\fP [\fI<even integer number>\fP]
.RS 4
Width of window borders. If the given width is an odd number, it will be incremented by 1.
.RE
.sp
\fBactive_window_border_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the border of the focused window.
.RE
.sp
\fBnormal_window_border_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the border of an unfocused window.
.RE
.sp
\fBinsert_feedback_color\fP [\fI0xAARRGGBB\fP]
.RS 4
Color of the \fBwindow \-\-insert\fP message selection.
.RE
.sp
\fBsplit_ratio\fP [\fI<FLOAT_SEL>\fP]
.RS 4
Default split ratio.
.RE
.sp
\fBauto_balance\fP [\fI<BOOL_SEL>\fP]
.RS 4
Balance the window tree upon change, so that all windows occupy the same area.
.RE
.sp
\fBmouse_modifier\fP [\fIcmd|alt|shift|ctrl|fn\fP]
.RS 4
Keyboard modifier used for moving and resizing windows.
.RE
.sp
\fBmouse_action1\fP [\fImove|resize\fP]
.RS 4
Action performed when pressing \fImouse_modifier\fP + \fIbutton1\fP.
.RE
.sp
\fBmouse_action2\fP [\fImove|resize\fP]
.RS 4
Action performed when pressing \fImouse_modifier\fP + \fIbutton2\fP.
.RE
.sp
\fBmouse_drop_action\fP [\fIswap|stack\fP]
.RS 4
Action performed when a bsp\-managed window is dropped in the center of some other bsp\-managed window.
.RE
.SS "Space Settings"
.sp
\fBlayout\fP [\fIbsp|stack|float\fP]
.RS 4
Set the layout of the selected space.
.RE
.sp
\fBtop_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the upper side of the selected space.
.RE
.sp
\fBbottom_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the lower side of the selected space.
.RE
.sp
\fBleft_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the left side of the selected space.
.RE
.sp
\fBright_padding\fP [\fI<integer number>\fP]
.RS 4
Padding added at the right side of the selected space.
.RE
.sp
\fBwindow_gap\fP [\fI<integer number>\fP]
.RS 4
Size of the gap that separates windows for the selected space.
.RE
.SS "Display"
.SS "General Syntax"
.sp
yabai \-m display [\fI<DISPLAY_SEL\fP>] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP \fI<DISPLAY_SEL>\fP
.RS 4
Focus the given display.
.RE
.SS "Space"
.SS "General Syntax"
.sp
yabai \-m space [\fI<SPACE_SEL>\fP] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP \fI<SPACE_SEL>\fP
.RS 4
Focus the given space.
.RE
.sp
\fB\-\-create\fP
.RS 4
Create a new space on the display of the selected space.
.RE
.sp
\fB\-\-destroy\fP
.RS 4
Remove the selected space.
.RE
.sp
\fB\-\-move\fP \fI<SPACE_SEL>\fP
.RS 4
Move position of the selected space to the position of the given space.
.br
The selected space and given space must both belong to the same display.
.RE
.sp
\fB\-\-swap\fP \fI<SPACE_SEL>\fP
.RS 4
Swap the selected space with the given space.
.br
The selected space and given space must both belong to the same display.
.RE
.sp
\fB\-\-display\fP \fI<DISPLAY_SEL>\fP
.RS 4
Send the selected space to the given display.
.RE
.sp
\fB\-\-balance\fP
.RS 4
Adjust the split ratios of the selected space so that all windows occupy the same area.
.RE
.sp
\fB\-\-mirror\fP \fIx\-axis|y\-axis\fP
.RS 4
Flip the tree of the selected space.
.RE
.sp
\fB\-\-rotate\fP \fI90|180|270\fP
.RS 4
Rotate the tree of the selected space.
.RE
.sp
\fB\-\-padding\fP \fIabs|rel:<top>:<bottom>:<left>:<right>\fP
.RS 4
Padding added at the sides of the selected space.
.RE
.sp
\fB\-\-gap\fP \fIabs|rel:<gap>\fP
.RS 4
Size of the gap that separates windows on the selected space.
.RE
.sp
\fB\-\-toggle\fP \fIpadding|gap|mission\-control|show\-desktop\fP
.RS 4
Toggle space setting on or off for the selected space.
.RE
.sp
\fB\-\-layout\fP \fIbsp|stack|float\fP
.RS 4
Set the layout of the selected space.
.RE
.sp
\fB\-\-label\fP [\fI<LABEL>\fP]
.RS 4
Label the selected space, allowing that label to be used as an alias in commands that take a \f(CRSPACE_SEL\fP parameter.
.br
If the command is called without an argument it will try to remove a previously assigned label.
.RE
.SS "Window"
.SS "General Syntax"
.sp
yabai \-m window [\fI<WINDOW_SEL>\fP] \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-focus\fP \fI<WINDOW_SEL>\fP
.RS 4
Focus the given window.
.RE
.sp
\fB\-\-swap\fP \fI<WINDOW_SEL>\fP
.RS 4
Swap position of the selected window and the given window.
.RE
.sp
\fB\-\-warp\fP \fI<WINDOW_SEL>\fP
.RS 4
Re\-insert the selected window, splitting the given window.
.RE
.sp
\fB\-\-stack\fP \fI<WINDOW_SEL>\fP
.RS 4
Stack the given window on top of the selected window.
.br
Any kind of warp operation performed on a stacked window will unstack it.
.RE
.sp
\fB\-\-insert\fP \fI<DIR_SEL>|stack\fP
.RS 4
Set the splitting mode of the selected window.
.br
If the current splitting mode matches the selected mode, the action will be undone.
.RE
.sp
\fB\-\-grid\fP \fI<rows>:<cols>:<start\-x>:<start\-y>:<width>:<height>\fP
.RS 4
Set the frame of the selected window based on a self\-defined grid.
.RE
.sp
\fB\-\-move\fP \fIabs|rel:<dx>:<dy>\fP
.RS 4
If type is \fIrel\fP the selected window is moved by \fIdx\fP pixels horizontally and \fIdy\fP pixels vertically, otherwise \fIdx\fP and \fIdy\fP will become its new position.
.RE
.sp
\fB\-\-resize\fP \fItop|left|bottom|right|top_left|top_right|bottom_right|bottom_left|abs:<dx>:<dy>\fP
.RS 4
Resize the selected window by moving the given handle \fIdx\fP pixels horizontally and \fIdy\fP pixels vertically. If handle is \fIabs\fP the new size will be \fIdx\fP width and \fIdy\fP height.
.RE
.sp
\fB\-\-ratio\fP \fIrel|abs:<dr>\fP
.RS 4
If type is \fIrel\fP the split ratio of the selected window is changed by \fIdr\fP, otherwise \fIdr\fP will become the new split ratio. A positive/negative delta will increase/decrease the size of the left\-child.
.RE
.sp
\fB\-\-toggle\fP \fIfloat|sticky|topmost|pip|shadow|border|split|zoom\-parent|zoom\-fullscreen|native\-fullscreen|expose\fP
.RS 4
Toggle the given property of the selected window.
.RE
.sp
\fB\-\-layer\fP \fI<LAYER>\fP
.RS 4
Set the stacking layer of the selected window.
.RE
.sp
\fB\-\-opacity\fP \fI<floating point number>\fP
.RS 4
Set the opacity of the selected window. The window will no longer be eligible for automatic change in opacity upon focus change.
.br
Specify the value \fI0.0\fP to reset back to full opacity OR have it be automatically managed through focus changes.
.RE
.sp
\fB\-\-display\fP \fI<DISPLAY_SEL>\fP
.RS 4
Send the selected window to the given display.
.RE
.sp
\fB\-\-space\fP \fI<SPACE_SEL>\fP
.RS 4
Send the selected window to the given space.
.RE
.sp
\fB\-\-minimize\fP
.RS 4
Minimizes the selected window. Only works on windows that provide a minimize button in its titlebar.
.RE
.sp
\fB\-\-deminimize\fP
.RS 4
Restores the selected window, if it is minimized. The window will only get focus if the owning application has focus.
.br
Note that you can also \fI\-\-focus\fP a minimized window to restore it as the focused window.
.RE
.sp
\fB\-\-close\fP
.RS 4
Closes the selected window. Only works on windows that provide a close button in its titlebar.
.RE
.SS "Query"
.SS "General Syntax"
.sp
yabai \-m query \fI<COMMAND>\fP [\fI<ARGUMENT>\fP]
.SS "COMMAND"
.sp
\fB\-\-displays\fP
.RS 4
Retrieve information about displays.
.RE
.sp
\fB\-\-spaces\fP
.RS 4
Retrieve information about spaces.
.RE
.sp
\fB\-\-windows\fP
.RS 4
Retrieve information about windows.
.RE
.SS "ARGUMENT"
.sp
\fB\-\-display\fP [\fI<DISPLAY_SEL>\fP]
.RS 4
Constrain matches to the selected display.
.RE
.sp
\fB\-\-space\fP [\fI<SPACE_SEL>\fP]
.RS 4
Constrain matches to the selected space.
.RE
.sp
\fB\-\-window\fP [\fI<WINDOW_SEL>\fP]
.RS 4
Constrain matches to the selected window.
.RE
.SS "Rule"
.sp
All registered rules that match the given filter will apply to a window in the order they were added.
.br
If multiple rules specify a value for the same argument, the latter rule will override that value as it was applied last.
.SS "General Syntax"
.sp
yabai \-m rule \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-add [\fI<ARGUMENT>\fP]\fP
.RS 4
Add a new rule.
.RE
.sp
\fB\-\-remove \fI<RULE_SEL>\fP\fP
.RS 4
Remove an existing rule with the given index or label.
.RE
.sp
\fB\-\-list\fP
.RS 4
Output list of registered rules.
.RE
.SS "ARGUMENT"
.sp
\fBlabel=\fI<LABEL>\fP\fP
.RS 4
Label used to identify the rule with a unique name
.RE
.sp
\fBapp[!]=\fI<REGEX>\fP\fP
.RS 4
Name of application. If \fI!\fP is present, invert the match.
.RE
.sp
\fBtitle[!]=\fI<REGEX>\fP\fP
.RS 4
Title of window. If \fI!\fP is present, invert the match.
.RE
.sp
\fBdisplay=\fI[^]<DISPLAY_SEL>\fP\fP
.RS 4
Send window to display. If \fI^\fP is present, follow focus.
.RE
.sp
\fBspace=\fI[^]<SPACE_SEL>\fP\fP
.RS 4
Send window to space. If \fI^\fP is present, follow focus.
.RE
.sp
\fBmanage=\fI<BOOL_SEL>\fP\fP
.RS 4
Window should be managed (tile vs float).
.RE
.sp
\fBsticky=\fI<BOOL_SEL>\fP\fP
.RS 4
Window appears on all spaces.
.RE
.sp
\fBlayer=\fI<LAYER>\fP\fP
.RS 4
Window is ordered within the given stacking layer.
.RE
.sp
\fBopacity=\fI<FLOAT_SEL>\fP\fP
.RS 4
Set window opacity. The window will no longer be eligible for automatic change in opacity upon focus change.
.RE
.sp
\fBborder=\fI<BOOL_SEL>\fP\fP
.RS 4
Window should draw border.
.RE
.sp
\fBnative\-fullscreen=\fI<BOOL_SEL>\fP\fP
.RS 4
Window should enter native macOS fullscreen mode.
.RE
.sp
\fBgrid=\fI<rows>:<cols>:<start\-x>:<start\-y>:<width>:<height>\fP\fP
.RS 4
Set window frame based on a self\-defined grid.
.RE
.SS "Signal"
.sp
A signal is a simple way for the user to react to some event that has been processed.
.br
Arguments are passed through environment variables.
.SS "General Syntax"
.sp
yabai \-m signal \fI<COMMAND>\fP
.SS "COMMAND"
.sp
\fB\-\-add event=\fI<EVENT>\fP action=\fI<ACTION>\fP [label=\fI<LABEL>\fP] [app[!]=\fI<REGEX>\fP] [title[!]=\fI<REGEX>\fP]\fP
.RS 4
Add an optionally labelled signal to execute an action after processing an event of the given type.
.br
Some signals can be specified to trigger based on the application name and/or window title.
.RE
.sp
\fB\-\-remove \fI<SIGNAL_SEL>\fP\fP
.RS 4
Remove an existing signal with the given index or label.
.RE
.sp
\fB\-\-list\fP
.RS 4
Output list of registered signals.
.RE
.SS "EVENT"
.sp
\fBapplication_launched\fP
.RS 4
Triggered when a new application is launched.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBapplication_terminated\fP
.RS 4
Triggered when an application is terminated.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBapplication_front_switched\fP
.RS 4
Triggered when the front\-most application changes.
.br
Passes two arguments: $YABAI_PROCESS_ID, $YABAI_RECENT_PROCESS_ID
.RE
.sp
\fBapplication_activated\fP
.RS 4
Triggered when an application is activated.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBapplication_deactivated\fP
.RS 4
Triggered when an application is deactivated.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBapplication_visible\fP
.RS 4
Triggered when an application is unhidden.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBapplication_hidden\fP
.RS 4
Triggered when an application is hidden.
.br
Eligible for \fBapp\fP filter.
.br
Passes one argument: $YABAI_PROCESS_ID
.RE
.sp
\fBwindow_created\fP
.RS 4
Triggered when a window is created.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_destroyed\fP
.RS 4
Triggered when a window is destroyed.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_focused\fP
.RS 4
Triggered when a window becomes the key\-window for its application.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_moved\fP
.RS 4
Triggered when a window changes position.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_resized\fP
.RS 4
Triggered when a window changes dimensions.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_minimized\fP
.RS 4
Triggered when a window has been minimized.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_deminimized\fP
.RS 4
Triggered when a window has been deminimized.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBwindow_title_changed\fP
.RS 4
Triggered when a window changes its title.
.br
Eligible for both \fBapp\fP and \fBtitle\fP filter.
.br
Passes one argument: $YABAI_WINDOW_ID
.RE
.sp
\fBspace_changed\fP
.RS 4
Triggered when the active space has changed.
.br
Passes two arguments: $YABAI_SPACE_ID, $YABAI_RECENT_SPACE_ID
.RE
.sp
\fBdisplay_added\fP
.RS 4
Triggered when a new display has been added.
.br
Passes one argument: $YABAI_DISPLAY_ID
.RE
.sp
\fBdisplay_removed\fP
.RS 4
Triggered when a display has been removed.
.br
Passes one argument: $YABAI_DISPLAY_ID
.RE
.sp
\fBdisplay_moved\fP
.RS 4
Triggered when a change has been made to display arrangement.
.br
Passes one argument: $YABAI_DISPLAY_ID
.RE
.sp
\fBdisplay_resized\fP
.RS 4
Triggered when a display has changed resolution.
.br
Passes one argument: $YABAI_DISPLAY_ID
.RE
.sp
\fBdisplay_changed\fP
.RS 4
Triggered when the active display has changed.
.br
Passes two arguments: $YABAI_DISPLAY_ID, $YABAI_RECENT_DISPLAY_ID
.RE
.sp
\fBmouse_down\fP
.RS 4
Triggered when a mouse button has been pressed.
.br
Passes two arguments: $YABAI_BUTTON, $YABAI_POINT
.RE
.sp
\fBmouse_up\fP
.RS 4
Triggered when a mouse button has been released.
.br
Passes two arguments: $YABAI_BUTTON, $YABAI_POINT
.RE
.sp
\fBmouse_dragged\fP
.RS 4
Triggered when the mouse is moved with one of its buttons pressed.
.br
Passes two arguments: $YABAI_BUTTON, $YABAI_POINT
.RE
.sp
\fBmouse_moved\fP
.RS 4
Triggered when the mouse is moved.
.br
Passes two arguments: $YABAI_BUTTON, $YABAI_POINT
.RE
.sp
\fBmission_control_enter\fP
.RS 4
Triggered when mission\-control activates.
.RE
.sp
\fBmission_control_check_for_exit\fP
.RS 4
Triggered periodically while mission\-control is active.
.RE
.sp
\fBmission_control_exit\fP
.RS 4
Triggered when mission\-control deactivates.
.RE
.sp
\fBdock_did_restart\fP
.RS 4
Triggered when Dock.app restarts.
.RE
.sp
\fBmenu_opened\fP
.RS 4
Triggered when a menu is opened.
.RE
.sp
\fBmenu_bar_hidden_changed\fP
.RS 4
Triggered when the macOS menubar \fIautohide\fP setting changes.
.RE
.sp
\fBdock_did_change_pref\fP
.RS 4
Triggered when the macOS Dock preferences changes.
.RE
.sp
\fBsystem_woke\fP
.RS 4
Triggered when macOS wakes from sleep.
.RE
.sp
\fBdaemon_message\fP
.RS 4
Triggered when yabai receives a message on its socket.
.RE
.SS "ACTION"
.sp
Arbitrary command executed through \fB/usr/bin/env sh \-c\fP
.SH "EXIT CODES"
.sp
If \fByabai\fP can\(cqt handle a message, it will return a non\-zero exit code.
.SH "AUTHOR"
.sp
Åsmund Vikane <aasvi93 at gmail.com>

src/yabai.c
#define SA_SOCKET_PATH_FMT      "/tmp/yabai-sa_%s.socket"

#define SOCKET_PATH_FMT         "/tmp/yabai_%s.socket"
#define LCFILE_PATH_FMT         "/tmp/yabai_%s.lock"

#define CLIENT_OPT_LONG         "--message"
#define CLIENT_OPT_SHRT         "-m"

#define DEBUG_VERBOSE_OPT_LONG  "--verbose"
#define DEBUG_VERBOSE_OPT_SHRT  "-V"
#define VERSION_OPT_LONG        "--version"
#define VERSION_OPT_SHRT        "-v"
#define CONFIG_OPT_LONG         "--config"
#define CONFIG_OPT_SHRT         "-c"

#define SCRPT_ADD_INSTALL_OPT   "--install-sa"
#define SCRPT_ADD_UNINSTALL_OPT "--uninstall-sa"
#define SCRPT_ADD_CHECK_OPT     "--check-sa"
#define SCRPT_ADD_LOAD_OPT      "--load-sa"

#define MAJOR 3
#define MINOR 3
#define PATCH 2

#define CONNECTION_CALLBACK(name) void name(uint32_t type, void *data, size_t data_length, void *context, int cid)
typedef CONNECTION_CALLBACK(connection_callback);
extern CGError SLSRegisterConnectionNotifyProc(int cid, connection_callback *handler, uint32_t event, void *context);

struct event_loop g_event_loop;
void *g_workspace_context;
struct process_manager g_process_manager;
struct display_manager g_display_manager;
struct space_manager g_space_manager;
struct window_manager g_window_manager;
struct mouse_state g_mouse_state;
struct event_tap g_event_tap;
struct daemon g_daemon;
int g_normal_window_level;
int g_floating_window_level;
int g_connection;

struct signal *g_signal_event[EVENT_TYPE_COUNT];
bool g_mission_control_active;
char g_sa_socket_file[MAXLEN];
char g_socket_file[MAXLEN];
char g_config_file[4096];
char g_lock_file[MAXLEN];
bool g_verbose;

static int client_send_message(int argc, char **argv)
{
    if (argc <= 1) {
        error("yabai-msg: no arguments given! abort..\n");
    }

    char *user = getenv("USER");
    if (!user) {
        error("yabai-msg: 'env USER' not set! abort..\n");
    }

    int sockfd;
    char socket_file[MAXLEN];
    snprintf(socket_file, sizeof(socket_file), SOCKET_PATH_FMT, user);

    if (!socket_connect_un(&sockfd, socket_file)) {
        error("yabai-msg: failed to connect to socket..\n");
    }

    int message_length = argc - 1;
    int argl[argc];

    for (int i = 1; i < argc; ++i) {
        argl[i] = strlen(argv[i]);
        message_length += argl[i];
    }

    char message[message_length];
    char *temp = message;

    for (int i = 1; i < argc; ++i) {
        memcpy(temp, argv[i], argl[i]);
        temp += argl[i];
        *temp++ = '\0';
    }

    if (!socket_write_bytes(sockfd, message, message_length)) {
        error("yabai-msg: failed to send data..\n");
    }

    shutdown(sockfd, SHUT_WR);

    int result = EXIT_SUCCESS;
    int byte_count = 0;
    char rsp[BUFSIZ];

    struct pollfd fds[] = {
        { sockfd, POLLIN, 0 }
    };

    while (poll(fds, 1, -1) > 0) {
        if (fds[0].revents & POLLIN) {
            if ((byte_count = recv(sockfd, rsp, sizeof(rsp)-1, 0)) <= 0) {
                break;
            }

            rsp[byte_count] = '\0';

            if (rsp[0] == FAILURE_MESSAGE[0]) {
                result = EXIT_FAILURE;
                fprintf(stderr, "%s", rsp + 1);
                fflush(stderr);
            } else {
                fprintf(stdout, "%s", rsp);
                fflush(stdout);
            }
        }
    }

    socket_close(sockfd);
    return result;
}

static void acquire_lockfile(void)
{
    int handle = open(g_lock_file, O_CREAT | O_WRONLY, 0600);
    if (handle == -1) {
        error("yabai: could not create lock-file! abort..\n");
    }

    struct flock lockfd = {
        .l_start  = 0,
        .l_len    = 0,
        .l_pid    = getpid(),
        .l_type   = F_WRLCK,
        .l_whence = SEEK_SET
    };

    if (fcntl(handle, F_SETLK, &lockfd) == -1) {
        error("yabai: could not acquire lock-file! abort..\n");
    }
}

static bool get_config_file(char *restrict filename, char *restrict buffer, int buffer_size)
{
    char *xdg_home = getenv("XDG_CONFIG_HOME");
    if (xdg_home && *xdg_home) {
        snprintf(buffer, buffer_size, "%s/yabai/%s", xdg_home, filename);
        if (file_exists(buffer)) return true;
    }

    char *home = getenv("HOME");
    if (!home) return false;

    snprintf(buffer, buffer_size, "%s/.config/yabai/%s", home, filename);
    if (file_exists(buffer)) return true;

    snprintf(buffer, buffer_size, "%s/.%s", home, filename);
    return file_exists(buffer);
}

static void exec_config_file(void)
{
    if (!*g_config_file && !get_config_file("yabairc", g_config_file, sizeof(g_config_file))) {
        notify("configuration", "could not locate config file..");
        return;
    }

    if (!file_exists(g_config_file)) {
        notify("configuration", "file '%s' does not exist..", g_config_file);
        return;
    }

    if (!ensure_executable_permission(g_config_file)) {
        notify("configuration", "could not set the executable permission bit for '%s'", g_config_file);
        return;
    }

    if (!fork_exec(g_config_file, NULL)) {
        notify("configuration", "failed to execute file '%s'", g_config_file);
        return;
    }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
static inline void init_misc_settings(void)
{
    char *user = getenv("USER");
    if (!user) {
        error("yabai: 'env USER' not set! abort..\n");
    }

    snprintf(g_sa_socket_file, sizeof(g_sa_socket_file), SA_SOCKET_PATH_FMT, user);
    snprintf(g_socket_file, sizeof(g_socket_file), SOCKET_PATH_FMT, user);
    snprintf(g_lock_file, sizeof(g_lock_file), LCFILE_PATH_FMT, user);

    NSApplicationLoad();
    signal(SIGCHLD, SIG_IGN);
    signal(SIGPIPE, SIG_IGN);
    CGSetLocalEventsSuppressionInterval(0.0f);
    CGEnableEventStateCombining(false);
    g_connection = SLSMainConnectionID();
    g_normal_window_level   = CGWindowLevelForKey(LAYER_NORMAL);
    g_floating_window_level = CGWindowLevelForKey(LAYER_ABOVE);
}
#pragma clang diagnostic pop

static CONNECTION_CALLBACK(connection_handler)
{
    event_loop_post(&g_event_loop, MISSION_CONTROL_ENTER, NULL, 0, NULL);
}

static void parse_arguments(int argc, char **argv)
{
    if ((string_equals(argv[1], VERSION_OPT_LONG)) ||
        (string_equals(argv[1], VERSION_OPT_SHRT))) {
        fprintf(stdout, "yabai-v%d.%d.%d\n", MAJOR, MINOR, PATCH);
        exit(EXIT_SUCCESS);
    }

    if ((string_equals(argv[1], CLIENT_OPT_LONG)) ||
        (string_equals(argv[1], CLIENT_OPT_SHRT))) {
        exit(client_send_message(argc-1, argv+1));
    }

    if (string_equals(argv[1], SCRPT_ADD_INSTALL_OPT)) {
        exit(scripting_addition_install());
    }

    if (string_equals(argv[1], SCRPT_ADD_UNINSTALL_OPT)) {
        exit(scripting_addition_uninstall());
    }

    if (string_equals(argv[1], SCRPT_ADD_CHECK_OPT)) {
        exit(scripting_addition_check());
    }

    if (string_equals(argv[1], SCRPT_ADD_LOAD_OPT)) {
        exit(scripting_addition_load());
    }

    for (int i = 1; i < argc; ++i) {
        char *opt = argv[i];

        if ((string_equals(opt, DEBUG_VERBOSE_OPT_LONG)) ||
            (string_equals(opt, DEBUG_VERBOSE_OPT_SHRT))) {
            g_verbose = true;
        } else if ((string_equals(opt, CONFIG_OPT_LONG)) ||
                   (string_equals(opt, CONFIG_OPT_SHRT))) {
            char *val = i < argc - 1 ? argv[++i] : NULL;
            if (!val) error("yabai: option '%s|%s' requires an argument!\n", CONFIG_OPT_LONG, CONFIG_OPT_SHRT);
            snprintf(g_config_file, sizeof(g_config_file), "%s", val);
        } else {
            error("yabai: '%s' is not a valid option!\n", opt);
        }
    }
}

int main(int argc, char **argv)
{
    if (argc > 1) {
        parse_arguments(argc, argv);
    }

    if (is_root()) {
        error("yabai: running as root is not allowed! abort..\n");
    }

    if (!ax_privilege()) {
        error("yabai: could not access accessibility features! abort..\n");
    }

    init_misc_settings();
    acquire_lockfile();

    if (!space_manager_has_separate_spaces()) {
        error("yabai: 'display has separate spaces' is disabled! abort..\n");
    }

    if (!event_loop_init(&g_event_loop)) {
        error("yabai: could not initialize event_loop! abort..\n");
    }

    process_manager_init(&g_process_manager);
    workspace_event_handler_init(&g_workspace_context);
    space_manager_init(&g_space_manager);
    window_manager_init(&g_window_manager);
    mouse_state_init(&g_mouse_state);

    event_loop_begin(&g_event_loop);
    display_manager_begin(&g_display_manager);
    space_manager_begin(&g_space_manager);
    window_manager_begin(&g_space_manager, &g_window_manager);
    process_manager_begin(&g_process_manager);
    workspace_event_handler_begin(&g_workspace_context);
    event_tap_begin(&g_event_tap, EVENT_MASK_MOUSE, mouse_handler);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1204, NULL);

    if (!socket_daemon_begin_un(&g_daemon, g_socket_file, message_handler)) {
        error("yabai: could not initialize daemon! abort..\n");
    }

    if (scripting_addition_is_installed()) {
        scripting_addition_load();
    } else {
        notify("scripting-addition", "payload is not installed, some features will not work!");
    }

    exec_config_file();
    CFRunLoopRun();
    return 0;
}
